File: frontend/lib/models/espacio.dart
/// RESERVIVES - Modelo de Espacio.


library;

import 'package:reservives/config/constants.dart';

enum TipoEspacio {
  pista('PISTA'),
  aula('AULA');

  final String value;
  const TipoEspacio(this.value);

  factory TipoEspacio.fromString(String value) {
    return TipoEspacio.values.firstWhere(
          (e) => e.value == value,
      orElse: () => TipoEspacio.pista,
    );
  }
}

class Espacio {
  final String id;
  final String nombre;
  final String? descripcion;
  final String? imagenUrl;
  final TipoEspacio tipo;
  final int precioTokens;
  final bool reservable;
  final bool requiereAutorizacion;
  final int antelacionDias;
  final String? ubicacion;
  final int? capacidad;
  final bool activo;
  final List<String> rolesPermitidos;
  final DateTime createdAt;
  final DateTime updatedAt;
  final bool isFavorite;

  const Espacio({
    required this.id,
    required this.nombre,
    this.descripcion,
    this.imagenUrl,
    required this.tipo,
    required this.precioTokens,
    required this.reservable,
    required this.requiereAutorizacion,
    required this.antelacionDias,
    this.ubicacion,
    this.capacidad,
    required this.activo,
    required this.rolesPermitidos,
    required this.createdAt,
    required this.updatedAt,
    this.isFavorite = false,
  });

  Espacio copyWith({
    String? id,
    String? nombre,
    String? descripcion,
    String? imagenUrl,
    TipoEspacio? tipo,
    int? precioTokens,
    bool? reservable,
    bool? requiereAutorizacion,
    int? antelacionDias,
    String? ubicacion,
    int? capacidad,
    bool? activo,
    List<String>? rolesPermitidos,
    DateTime? createdAt,
    DateTime? updatedAt,
    bool? isFavorite,
  }) {
    return Espacio(
      id: id ?? this.id,
      nombre: nombre ?? this.nombre,
      descripcion: descripcion ?? this.descripcion,
      imagenUrl: imagenUrl ?? this.imagenUrl,
      tipo: tipo ?? this.tipo,
      precioTokens: precioTokens ?? this.precioTokens,
      reservable: reservable ?? this.reservable,
      requiereAutorizacion: requiereAutorizacion ?? this.requiereAutorizacion,
      antelacionDias: antelacionDias ?? this.antelacionDias,
      ubicacion: ubicacion ?? this.ubicacion,
      capacidad: capacidad ?? this.capacidad,
      activo: activo ?? this.activo,
      rolesPermitidos: rolesPermitidos ?? this.rolesPermitidos,
      createdAt: createdAt ?? this.createdAt,
      updatedAt: updatedAt ?? this.updatedAt,
      isFavorite: isFavorite ?? this.isFavorite,
    );
  }

  factory Espacio.fromJson(Map<String, dynamic> json) {
    return Espacio(
      id: json['id'] as String,
      nombre: json['nombre'] as String,
      descripcion: json['descripcion'] as String?,
      imagenUrl: AppConstants.resolveApiUrl(json['imagen_url'] as String?),
      tipo: TipoEspacio.fromString(json['tipo'] as String),
      precioTokens: json['precio_tokens'] as int,
      reservable: json['reservable'] as bool,
      requiereAutorizacion: json['requiere_autorizacion'] as bool,
      antelacionDias: json['antelacion_dias'] as int,
      ubicacion: json['ubicacion'] as String?,
      capacidad: json['capacidad'] as int?,
      activo: json['activo'] as bool,
      rolesPermitidos: List<String>.from(json['roles_permitidos'] ?? []),
      createdAt: DateTime.parse(json['created_at'] as String),
      updatedAt: DateTime.parse(json['updated_at'] as String),
      isFavorite: json['is_favorite'] as bool? ?? false,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'nombre': nombre,
      'descripcion': descripcion,
      'imagen_url': imagenUrl,
      'tipo': tipo.value,
      'precio_tokens': precioTokens,
      'reservable': reservable,
      'requiere_autorizacion': requiereAutorizacion,
      'antelacion_dias': antelacionDias,
      'ubicacion': ubicacion,
      'capacidad': capacidad,
      'activo': activo,
      'roles_permitidos': rolesPermitidos,
      'created_at': createdAt.toIso8601String(),
      'updated_at': updatedAt.toIso8601String(),
      'is_favorite': isFavorite,
    };
  }
}

File: backend/app/models/espacio.py
"""
Modelo de Espacio.

Representa los espacios reservables del instituto (pistas deportivas y aulas).
También incluye la tabla intermedia de roles permitidos por espacio.
"""

import enum
import uuid
from datetime import datetime

from sqlalchemy import Boolean, DateTime, Enum, ForeignKey, Integer, String, Text, func
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.database import Base


class TipoEspacio(str, enum.Enum):
    """Tipos de espacio disponibles."""
    PISTA = "PISTA"
    AULA = "AULA"


class Espacio(Base):
    """Modelo SQLAlchemy para la tabla 'espacios'."""
    __tablename__ = "espacios"

    id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), primary_key=True, default=uuid.uuid4
    )
    nombre: Mapped[str] = mapped_column(String(150), nullable=False)
    descripcion: Mapped[str | None] = mapped_column(Text)
    imagen_url: Mapped[str | None] = mapped_column(String(500))
    tipo: Mapped[TipoEspacio] = mapped_column(
        Enum(TipoEspacio, name="tipo_espacio", create_type=False),
        nullable=False
    )
    precio_tokens: Mapped[int] = mapped_column(Integer, nullable=False, default=0)
    reservable: Mapped[bool] = mapped_column(Boolean, nullable=False, default=True)
    requiere_autorizacion: Mapped[bool] = mapped_column(Boolean, nullable=False, default=False)
    antelacion_dias: Mapped[int] = mapped_column(Integer, nullable=False, default=7)
    ubicacion: Mapped[str | None] = mapped_column(String(200))
    capacidad: Mapped[int | None] = mapped_column(Integer)
    activo: Mapped[bool] = mapped_column(Boolean, nullable=False, default=True)
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now()
    )
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now(), onupdate=func.now()
    )

    # Relaciones
    roles_permitidos = relationship(
        "EspacioRolPermitido", back_populates="espacio",
        lazy="selectin", cascade="all, delete-orphan"
    )
    reservas = relationship("ReservaEspacio", back_populates="espacio", lazy="selectin")
    tramos_permitidos = relationship(
        "EspacioTramoPermitido", back_populates="espacio",
        lazy="selectin", cascade="all, delete-orphan"
    )

    def __repr__(self) -> str:
        return f"<Espacio {self.nombre} ({self.tipo.value})>"


class EspacioRolPermitido(Base):
    """Tabla intermedia: qué roles pueden reservar cada espacio."""
    __tablename__ = "espacio_rol_permitido"

    id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), primary_key=True, default=uuid.uuid4
    )
    espacio_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("espacios.id", ondelete="CASCADE"), nullable=False
    )
    rol: Mapped[str] = mapped_column(
        Enum("ALUMNO", "PROFESOR", "ADMIN", name="rol_usuario", create_type=False),
        nullable=False
    )

    # Relación inversa
    espacio = relationship("Espacio", back_populates="roles_permitidos")